### ISSUE TITLE
Documentation feedback

### ISSUE THREAD

**Issue by ibroom** [Fliplet/fliplet-agent#2]
1. can we update the example to be sqlserver and odbc based as I think this is all we support at launch?
2. why doesn't the software come with a demo config file that I can just copy, or why can't I generate a config file by running DIS?
3. in the config can we include links to the relevant docs e.g. API token https://developers.fliplet.com/REST-API/authenticate.html 2nd section (I think the 'how to make a token' should be first in this doc btw)
4. database drivers - we need to list the options
5. under the sql only options, you say 'remove these lines if not necessary' should we instead comment them by default and people can uncomment them?
6. crontab syntax - as the target is windows users I suggest a link to docs about crontab and a few options so that the dev can just uncomment them if they want them instead of having to read the docs e.g. 15 min, 1 hour, 12 hour, 1 day, midnight only working days. I'd also state that 'When testing, if you have init sync enabled your agent will sync as soon as it is run so restarting the agent is the fastest way to test if the configuration is working'
7. The select syntax needs to explain 'the column names must match the data source or new columns will be added, use 'as' to map database columns to Fliplet data source column names and avoid new columns being created'
8. I think the data source ID and the Fliplet token should be next to each other at the top of the doc, we need instructions how to get the data source ID as many DIS users will not be Studio users
9. The test config option should be listed before we tell someone to run their script, currently it is after
10. Advanced use section - as this is using libraries, I recommend we link to the libraries and their docs in this section as I have no idea how to use this without understanding what libraries are being called to process the config e.g. X is used to connect to database Y is used to parse the data, etc
11. What if I have multiple config yml files, is this supported and how do I run them? When we first discussed it, I had assumed the config files would all be in a single folder and all the config files would be read and run on load by a single node thread but it looks like I have to run a separate node thread (therefore a separate service) for each config file? Can I use 1 agent with multiple config files? Can we include an example of how to start multiple config files?
12. Did we include a hook method that developers can use to write custom JS code to parse their code? I don't think it's necessary but I wasn't sure if it was available after reading the advanced mode. If it isn't, why is advanced mode useful?

**Comment by squallstar** [comment]
1. done

2. It does come with a sample file that you can copy, check the screenshot below:

![image](https://user-images.githubusercontent.com/574210/47156357-40214280-d2e7-11e8-933a-6251ecb96f9a.png)

You have two ways to get started. Generating a file via the software is possible but it's more unplanned work which gives low value.

3. done

4. done

5. done

6. done

7. done

8. done

9. done

10. partially fixed, let's discuss on a call

11. Advanced mode (via Javascript config file) supports setting up multiple synchronisations within the same file, while basic mode currently don't. This is out of scope for our release because you can simply run multiple services. This is not an issue because each service will then have its own STDOUT/STDERR (the log files streamed to the OS).

If you believe this should be part of phase 1, let's please discuss on a call.

12. What do you mean by "a hook to write custom JS code to parse their code"? What code does need to be parsed? If you're talking about code to run before/after queries and for mapping the data then you deprioritised that for phase 4 of the work: https://docs.google.com/spreadsheets/d/1JkBhlG5fj914BcOMstagvwDQtvSj30Mc-4EvZlNTdIQ/edit#gid=710251963

**Comment by ibroom** [comment]
Is this an alternative - YML nesting? Is this hard for you to parse?

job name: sync users
  enter all the config here
job name: sync news
  enter all the config here, may duplicate the db, token, etc details but
it keeps the job processing easy

Actually we could have this to reuse the same DB connection but switch DB

driver: mssql
username: blah
password: abc123

job name: sync news
  database: news
  sql: 'select ID from...'
  primary key: id
  data source id: 1234
  token: eu-...

job name: sync users
  database: people
  sql: 'select ID from...'
  primary key: id
  data source id: 1234
  token: eu-...

**Comment by squallstar** [comment]
@ibroom yes, that can be done although it does introduce some risky bits like being aware that tabs/spaces do matter in YML. The reason on why I didn't suggest it first is because (as I thought in first place) clients might simply use advanced mode to get lots more features and granular control over the data.

**Comment by ibroom** [comment]
Can we validate the YML first? If so this reduces the risk.

Alternatively could do something like vhosts in apache that are explicitly
grouped by a wrapping statement and avoids tabs/spaces to control sections?

**Comment by squallstar** [comment]
@ibroom we do validate the YML, I'm good with defining multiple operations as long as clients understands how operations must be defined in the YML. E.g.:

```yml
# Shared settings
auth_token: ...
database_host: ...

# List of operations below

'My first operation':
  query: 'select...'
  primary_column: 'id'
  frequency: '...'

'My second operation':
  query: 'select...'
  primary_column: 'id'
  frequency: '...'
```

**Comment by ibroom** [comment]
I think this makes sense. For indented sections, is there an end block
parameter we can use instead of relying on indents? I don't think this is
the standard but it protects us against mistakes:

'My second operation':
  query: 'select...'
  primary_column: 'id'
  frequency: '...'

End

**Comment by squallstar** [comment]
@ibroom no, there's no such thing like an end block on YML since it relies on indentation as I stated in my previous messages, hence why my concern on the fact that people setting up this might simply make mistakes in the indentation and get errors. However, if the validation is robust enough I don't see any problem with it.

Linking here refs of YAML official spec http://yaml.org/spec/1.2/spec.html

**Comment by ibroom** [comment]
Thanks Nick, I agree validation and hopefully testing should be sufficient.
